okhttp/src/main/java/okhttp3/Body.java

```diff
@@ -0,0 +1,263 @@
+import java.io.Reader;
+import java.nio.charset.Charset;
+import okhttp3.internal.Util;
+import okhttp3.internal.http.HttpHeaders;
+import okio.Buffer;
+import okio.BufferedSink;
+import okio.BufferedSource;
+import okio.ByteString;
+import okio.Okio;
+
+import static okhttp3.internal.Util.UTF_8;
+
+/**
+ * The body for HTTP requests or responses.
+ */
+public class Body implements Closeable {
+  public static final byte[] EMPTY_ARRAY = new byte[0];
+  public static final Body EMPTY = new Body(null, EMPTY_ARRAY);
+
+  protected final MediaType contentType;
+  protected final long contentLength;
+  protected final BufferedSource readableContent;
+  protected final byte[] writableContent;
+
+  private Reader reader;
+
+  public Body() {
+    this(null, -1, null, null);
+  }
+
+  public Body(MediaType contentType) {
+    this(contentType, -1, null, null);
+  }
+
+  public Body(MediaType contentType, byte[] content) {
+    this(contentType, content.length, content, new Buffer().write(content));
+  }
+
+  public Body(MediaType contentType, Buffer content) {
+    this(contentType, content.size(), content.readByteArray(), content);
+  }
+
+  public Body(MediaType contentType, BufferedSource content) {
+    this(contentType, -1, null, content);
+  }
+
+  public Body(MediaType contentType, long contentLength, BufferedSource content) {
+    this(contentType, contentLength, null, content);
+  }
+
+  public Body(MediaType contentType, long contentLength, byte[] writableContent,
+      BufferedSource readableContent) {
+    this.contentType = contentType;
+    this.contentLength = contentLength;
+    this.writableContent = writableContent;
+    this.readableContent = readableContent;
+  }
+
+  public MediaType contentType() {
+    return contentType;
+  }
+
+  public void writeTo(BufferedSink sink) throws IOException {
+    if (writableContent == null) throw new UnsupportedOperationException("Not writable");
+    sink.writeAll(new Buffer().write(writableContent));
+  }
+
+  /**
+   * Returns the content length or -1 if the content length is unknown.
+   */
+  public long contentLength() throws IOException {
+    return contentLength;
+  }
+
+  public BufferedSource source() {
+    if (readableContent == null) throw new UnsupportedOperationException("Not readable");
+    return readableContent;
+  }
+
+  @Override public void close() {
+    Util.closeQuietly(source());
+  }
+
+  /**
+   * Returns the body as a byte array.
+   *
+   * <p>This method loads entire body into memory. If the body is very large this may trigger
+   * an {@link OutOfMemoryError}. Prefer to stream the body if this is a possibility
+   * for your request or response.
+   */
+  public final byte[] bytes() throws IOException {
+    long contentLength = contentLength();
+    if (contentLength > Integer.MAX_VALUE) {
+      throw new IOException("Cannot buffer entire body for content length: " + contentLength);
+    }
+
+    byte[] bytes;
+    try {
+      bytes = source().readByteArray();
```

Comment:
This method requires a readable source but could support writeable bodies by writing to `new Buffer()` and then calling `readByteArray()`.